src/Castle.Facilities.AspNetCore/ICastleWindsorMiddleware.cs

```diff
@@ -0,0 +1,26 @@
+﻿// Copyright 2004-2018 Castle Project - http://www.castleproject.org/
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle.Facilities.AspNetCore
+{
+	using System;
+	using System.Threading.Tasks;
+
+	using Microsoft.AspNetCore.Http;
+
+	public interface ICastleWindsorMiddleware
```

Comment:
Could you explain why this couldn't use `IMiddleware`? Does ASP.NET Core automatically wire up `IMiddleware`s not allowing them to be resolved from the container?

It looks like ASP.NET Core already handles this, see the ["Per-request dependencies" section](
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware?tabs=aspnetcore2x), is this functionality useful?

Reply:
> Could you explain why this couldn't use IMiddleware? Does ASP.NET Core automatically wire up IMiddlewares not allowing them to be resolved from the container?

Great question. You can definitely do this but you are moving away from the Windsor container and falling back onto "Framework Configured Container". This creates problems if your middleware consumes dependencies that are know to Windsor but not registered within the IServiceCollection. If anyone decided to go down this route they will end up with a mish mash of registrations across both containers. This should be avoided at all costs. 

> It looks like ASP.NET Core already handles this, see the "Per-request dependencies" section, is this functionality useful?

If their middleware does not consume any dependencies, then this functionality is useful. Otherwise it is recommended you use our extension because then you are free to consume your own types(registered in Windsor) and framework types which is dealt with in the sub resolvers.

I did update the XML documentation for this extension.

Reply:
This still feels icky as it means your application code has to implement a container specific interface otherwise the container won't let you use it. It feels a bit like the lifestyle attributes Windsor has that aren't really recommended as registration/configuration should happen outside the application code, or the anti-pattern that is depending on `IKernel` from non-factory components.

If a user did implement `IMiddleware` or used someone else's middleware would ASP.NET Core automatically wire it up, or can we just tell them not to register it with ASP.NET Core and instead register it with the Windsor facility?

I think that the type name (`Castle.Facilities.AspNetCore.ICastleWindsorMiddleware`) is something that keeps sticking its nose at me. Having the product name in the interface name is something I don't believe we've ever had to do before which sort of tells me we are doing something wrong, especially that Windsor doesn't have middleware but that this is a workaround.

> I did update the XML documentation for this extension.

FYI, the interface doesn't have any documentation?

Reply:
> This still feels icky as it means your application code has to implement a container specific interface otherwise the container won't let you use it.

You are right it is superfluous and can be removed altogether. 

 > If a user did implement IMiddleware or used someone else's middleware would ASP.NET Core automatically wire it up, or can we just tell them not to register it with ASP.NET Core and instead register it with the Windsor facility?

Not aware of anything that wires this up for you. 

For framework middleware:

 - [You still have to "tell" the IServiceCollection the middleware exists as a resolvable type](https://github.com/fir3pho3nixx/Windsor.AspNetCore/commit/15f5833a4b097f825f223cc788e3564bbdc88dd4#diff-68d907e65062d369b9f82e62dd896c6bR38).

 - [You have to register the same type as middleware with the app builder](https://github.com/fir3pho3nixx/Windsor.AspNetCore/commit/15f5833a4b097f825f223cc788e3564bbdc88dd4#diff-68d907e65062d369b9f82e62dd896c6bR54).  

 - It does not resolve types registered in Windsor, if you need this then register it in your container and use [this extension method instead](https://github.com/fir3pho3nixx/Windsor.AspNetCore/commit/15f5833a4b097f825f223cc788e3564bbdc88dd4#diff-573837522f7d6c1d750af3ef541e4629R70).

 > I think that the type name (Castle.Facilities.AspNetCore.ICastleWindsorMiddleware) is something that keeps sticking its nose at me. 

Let's delete it. We still need the extension method though right?

Reply:
Yay, `ICastleWindsorMiddleware` is gone 🎉 .

Yes, we'll still need the extension method, you were saying trying to do it via the framework container will just cause too much pain so happy for it to stay, however I noticed you renamed the `IApplicationBuilder` extension method:

```
app.UseCastleWindsorMiddleware<CustomMiddleware>(container);
... to ...
app.UseResolvableMiddleware<CustomMiddleware>(container);
```

I wonder if this new name is very clear, if you didn't want "Windsor" in the extension method name (which I thought was convention for these IApplicationBuilder thingys) then naming it just `UseMiddleware` like the framework one would explain just as much but you'd have an overload with the container.

```
app.UseMiddleware<CustomMiddleware>(); // register with framework
app.UseMiddleware<CustomMiddleware>(container); // register with Windsor
```

Or maybe name it `app.UseMiddlewareViaWindsor<CustomMiddleware>(container)`?